include/unifex/create.hpp

```diff
@@ -36,7 +36,7 @@ struct _op {
     template (typename... Ts)
       (requires receiver_of<Receiver, Ts...>)
     void set_value(Ts&&... ts) noexcept(is_nothrow_receiver_of_v<Receiver, Ts...>) {
-      unifex::set_value((Receiver&&) rec_, (Ts&&) ts...);
+      unifex::set_value((Receiver&&) rec_, (ValueTypes) ts...);
```

Comment:
The conversion from `Ts&&` to `ValueTypes` has me a bit squirrely.  The whole point of this change is to preserve the cvref qualifiers on the type parameters to `create<>()` so we obviously need to perform this conversion anytime `set_value` is invoked with inexactly-typed parameters, but I'm not sure I fully understand all the corner cases.

I think it's right for the function to take arbitrary forwarding references (rather than, say, `ValueTypes&&...`) so that:
 - natural implicit conversions can happen (e.g. a `create<double>` that is completed with `set_value(0)` ought to work), and
 - any exceptions that arise happen _inside_ `set_value` so that we can capture-and-forward them automatically.

But, I worry that we'll copy or move more often than necessary if we get the details wrong.

Maybe the solution is to factor the conversion into a functor that performs the operation that `convertible_to` checks for?

```
// defined at namespace scope
template <typename From, typename To, typename = void>
struct do_convert {
  // default to no conversion
  To operator()(From&&) = delete;
};

template <typename From>
struct do_convert<From, From, void> {
  // don't construct anything when "conversion" is the identity
  From&& operator()(From&& obj) noexcept {
    return static_cast<From&&>(obj);
  }
};

template <typename From, typename To>
struct do_convert<From, To, std::enable_if_t<convertible_to<From, To>>> {
  // explicitly perform the operation that we've checked is implicitly allowed
  To operator()(From&& obj) {
    return static_cast<From&&>(obj);
  }
};

template(typename T...) //
    (requires (convertible_to<T, ValueTypes> && ...)) //
void set_value(T&&... values) noexcept {
  UNIFEX_TRY {
    unifex::set_value(std::move(rec_), do_convert<T, ValueTypes>{}(static_cast<T&&>(values))...);
  }
  UNIFEX_CATCH(...) {
    unifex::set_error(std::move(rec_), std::current_exception());
  }
}
```

The part of the above I'm most uncertain about is the type parameters given to `do_convert`. Should it be
 - `do_convert<T, ValueTypes>`,
 - `do_convert<T&&, ValueTypes>`,
 - `do_convert<T, ValueTypes&&>`, or
 - `do_convert<T&&, ValueTypes&&>`?

For `create<cv int&>` and `set_value` invoked with a cv `int&`, they're all equivalent because of reference collapsing and they'll all pick the identity, which avoids unnecessary copies/moves and correctly forwards "the same" `int` from the `set_value` call site through to the _Receiver's_ `set_value`.

For `create<cv int&>` and `set_value` invoked with an rvalue `int` or with anything convertible to an `int`, like a cvref-qualified `double`, all of the above options will find the default template expansion and (correctly) block the conversion.

How do `create<cv int>` and `create<cv int&&>` differ?  There _is_ a difference between
```
std::unique_ptr<int> foo() {
  return std::move(*this).foo_;
}
```
and
```
std::unique_ptr<int>&& foo() {
  return std::move(*this).foo_;
}
```
For one thing `foo_` is unconditionally left null in the first one, but the second one is conditional, depending on what the caller does with the result.

Perhaps the second type argument to `do_convert` should be `ValueTypes` and not `ValueTypes&&` because it would allow us to distinguish between `create<int>` and `create<int&&>`.

To get the unconditional move behaviour for `create<int>` and the conditional move behaviour for `create<int&&>`, I think we need the first type argument to `do_convert` to be `T&&`; that will choose the identity "conversion" (and thus maintain reference stability) when `ValueTypes` is an rvalue reference, and pick the `T&& -> T` conversion when `ValueTypes` is not a reference at all.

So that leaves us with
```
unifex::set_value(std::move(rec_), do_convert<T&&, ValueTypes>{}(static_cast<T&&>(values))...);
```
Does my reasoning seem sound?

If so, is `do_convert` even necessary?  Maybe I've just complicated a cast expression.

Should this be
```suggestion
      UNIFEX_TRY {
        unifex::set_value(std::move(rec_), ValueTypes{static_cast<Ts&&>(ts)}...);
      }
      UNIFEX_CATCH(...) {
        unifex::set_error(std::move(rec_), std::current_exception());
      }
```
with a `requires (convertible_to<Ts, ValueTypes> && ...)` clause on the method?

Reply:
> But, I worry that we'll copy or move more often than necessary if we get the details wrong.

I had a similar concern. My original logic of `set_value((Receiver&&) rec_, (ValueTypes) ts...)` definitely incurs unnecessary copies or moves (I've added tests in my latest commit to capture this number of copies or moves to ensure no unnecessary copies or moves are done).

I'm having a hard time understanding whether `do_convert<Ts&&, ValueTypes>` or `do_convert<Ts, ValueTypes>` is right. Consider calling `set_value<AnObject>` on `create<AnObject>`. I think this will not hit the `do_convert<From, From, void>` specialization if `do_convert<Ts&&, ValueTypes>` is used and instead will end up calling the other which leads to a copy of the Object. In the latest commit I pushed, my new tests pass with both versions, so I don't think I have a test that exposes the right condition where `Ts&&` vs `Ts` in `do_convert` matters. I did add a test for the "conditional move" behavior (`create<TrackingObject&&>`).

|Ts|ValueTypes|Desired Result passed to the inner set_value|do_convert<Ts, ValueTypes>|do_convert<Ts&&, ValueTypes>|
|-|-|-|-|-|
|int|int|int&&|int&&|int|
|int&|int|int|int|int|

Reply:
Note, I added `Co-authored-by: Ian Petersen <ispeters@fb.com>` in the commit where I used `do_convert`. LMK if you are OK with this.

Reply:
> I'm having a hard time understanding whether `do_convert<Ts&&, ValueTypes>` or `do_convert<Ts, ValueTypes>` is right. Consider calling `set_value<AnObject>` on `create<AnObject>`. I think this will not hit the `do_convert<From, From, void>` specialization if `do_convert<Ts&&, ValueTypes>` is used and instead will end up calling the other which leads to a copy of the Object.

I'm currently leaning towards "`create<AnObject>` ought to always construct a new object" (whether by copy or by move is up to the value category of the argument to `set_value`).  In other words, `do_convert<Ts&&, ValuteTypes>` _not_ matching `do_convert<From, From, void>` is, I think, the right choice. If there are scenarios where `create<AnObject>` _doesn't_ create a new object then I think it's indistinguishable from `create<AnObject&&>`, right?

If I think of `set_value` as "how _Sender/Receiver_ implements `return`" then requiring `create<AnObject>` to always create a new object is akin to saying "_Sender/Receiver_ doesn't have RVO," which I think is just true.

I think there are two choices here with different trade-offs:
1. `create<AnObject>` always creates a new object and is distinguishable from `create<AnObject&&>`, or
2. `create<AnObject>` is the same as `create<AnObject&&>` and the algorithm never constructs a new object for you except when performing an implicit conversion.

I think the danger of 1 is unintentional copy/move, which can be "fixed" by changing to `create<AnObject&&>` and the danger of 2 is unintentional aliasing, which can be "fixed" by deliberately passing a temporary to `set_value` (i.e. `set_value(AnObject{anObject})`) or by receiving the value(s) by value (e.g. `| then([](auto a) { return a; })`).

I feel like the danger of 1 is less dangerous than the danger of 2 and the cost of safety isn't that high, but what do you think?

Also, if we make `create<AnObject>` and `create<AnObject&&>` the same, have we thrown away the ability to send values in the pursuit of the ability to send references?

Reply:
Re: giving me co-authorship credit, that's kind of you.  I feel like I'm playing the role of editor here, not co-author, so it feels a bit like undue credit, but I don't what the open source conventions are around this kind of thing.

Reply:
Ok, now I see - I think `create<AnObject>` should construct an object, that seems like the right semantics for "how Sender/Receiver implement `return`." Thanks for explaining the key difference. I added a test `CreateObjectLeadsToNewObject` which only works with `do_convert<Ts&&, ValueTypes>`, which ensures that the move constructor is invoked at least once (it's never invoked with `do_convert<Ts, ValueTypes>`).

Re: credit - im not sure of the convention either. Not a big deal, and fair point about being "editor," so I rebased and removed it.